### PR TITLE
Restructure kassert

### PR DESCRIPTION
**Pull request by DanielSeemaier**
* Move expression decomposition to its own file 
* Move logger to its own file
* Move macro implementation helpers to their own file 
* Remove KaMPI.ng specific assertion levels
* Multiply predefined assertion levels by 10 to allow more levels in between 

Code didn't change, just moved around declarations

### CODE REVIEW

File: include/kassert/internal/logger.hpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+#pragma once
+
+#include <ostream>
+#include <sstream>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace kassert {
```
Comment (lukashuebner) [inline include/kassert/internal/logger.hpp:25]: Maybe `namespace kassert::internal {`?

File: include/kassert/internal/logger.hpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+/// @brief Determines whether a value of type \c ValueT can be streamed into an output stream of type \c StreamT.
+/// @ingroup expression-expansion
+/// @tparam StreamT An output stream overloading the \c << operator.
+/// @tparam ValueT A value type that may or may not be used with \c StreamT::operator<<.
+template <typename StreamT, typename ValueT>
+constexpr bool is_streamable_type = is_streamable_type_impl<StreamT, ValueT>::value;
+} // namespace internal
+
+/// @brief Simple wrapper for output streams that is used to stringify values in assertions and exceptions.
```
Comment (lukashuebner) [inline include/kassert/internal/logger.hpp:44]: `namespace kassert {`

File: include/kassert/internal/logger.hpp [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    StreamT&& stream() {
+        return std::forward<StreamT>(_out);
+    }
+
+private:
+    /// @brief The underlying streaming object.
+    StreamT&& _out;
+};
+
+namespace internal {
```
Comment (lukashuebner) [inline include/kassert/internal/logger.hpp:80]: `namespace kassert::internal`, and so on. I think it's more readable but I don't want to insist on this. :-)

File: include/kassert/internal/macros.hpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
+// This file is part of KaMPI.ng.
```
Comment (lukashuebner) [inline include/kassert/internal/macros.hpp:1]: Should we name the file `assertion_macro.hpp`?

File: include/kassert/kassert.hpp [hunk 5]
```diff
@@ -245,451 +122,31 @@ class KassertException : public std::exception {
...
-namespace assert {
+namespace kassert::assert {
 /// @defgroup assertion-levels Assertion levels
 /// Predefined assertion levels.
 ///
 /// @{
 
 /// @brief Assertion level for exceptions if exception mode is disabled.
-#define KASSERT_ASSERTION_LEVEL_KTHROW 1
+#define KASSERT_ASSERTION_LEVEL_KTHROW 10
```
Comment (lukashuebner) [inline include/kassert/kassert.hpp:135]: You're combining multiple changes here, but I guess it's fine.

**Review by lukashuebner — APPROVED** [review]
Okay, so I did not read every line. Assuming you did not change anything except:

* Copying lines from `kassert.hpp` to the tree new files.
* Multiply all assertions levels by 10.

this change is fine by me. Treat my comments as suggestions :-)

File: include/kassert/internal/logger.hpp (hunk 1 above)
Comment (DanielSeemaier) [reply include/kassert/internal/logger.hpp:25]: The idea here is that the user should overload `Logger::operator<<` for his own data types. Hence he could come into contact with this class, which is why it's not placed in `kamping::internal`

**Comment by lukashuebner** [comment]
Oh, I just read: "Remove KaMPI.ng specific assertion levels". So the new assertion levels are:

#define KASSERT_ASSERTION_LEVEL_KTHROW 10
#define KASSERT_ASSERTION_LEVEL_NORMAL 30

So we explicitly removed the LIGHT, HEAVY, and communication levels? Why?

**Comment by DanielSeemaier** [comment]
> Oh, I just read: "Remove KaMPI.ng specific assertion levels". So the new assertion levels are:
> 
> #define KASSERT_ASSERTION_LEVEL_KTHROW 10 #define KASSERT_ASSERTION_LEVEL_NORMAL 30
> 
> So we explicitly removed the LIGHT, HEAVY, and communication levels? Why?

My thinking was that the Kassert library only defines the assertion levels that it uses internally, i.e., normal and for exceptions. The user (i.e., KaMPI.ng) is free to declare more assertions.

**Comment by lukashuebner** [comment]
So the user would then define all the assertion levels he uses himself? I.e. we get the `KAMPING_ASSERTION_LEVEL_` prefix back? Sounds good, but we could provide an example file containing some assertion definitions – if you have the spare time. We'd still need the assertions levels as preprocessor defines and not as constexpr in a namespace, right? Else we cannot use them to disable code which uses variables which are non-existent on the current assertion level.

**Comment by DanielSeemaier** [comment]
> So the user would then define all the assertion levels he uses himself? I.e. we get the `KAMPING_ASSERTION_LEVEL_` prefix back? 

yes exactly 

> Sounds good, but we could provide an example file containing some assertion definitions – if you have the spare time. 

I've added it to the README. I'm not happy with the way the library gets currently configured (magic constant 30), but that should probably be changed in another PR. 

> We'd still need the assertions levels as preprocessor defines and not as constexpr in a namespace, right? Else we cannot use them to disable code which uses variables which are non-existent on the current assertion level.

yes correct

**Pull request merged**